lib/update/cluster/phases/bootstrap.go

```diff
@@ -18,25 +18,233 @@ package phases
+	executor.existingEnviron = env.GetKeyValues()
+	return &executor, nil
+}
+
+// Execute executes the bootstrap phase locally, e.g. exports new gravity
+// binary, creates new secrets/config packages in the local backend and
+// initializes local operation state
+func (p *updatePhaseBootstrapLeader) Execute(ctx context.Context) error {
+	err := p.rotateConfigAndSecrets()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return p.bootstrap.Execute(ctx)
+}
+
+// Rollback is no-op for this phase
+func (p *updatePhaseBootstrapLeader) Rollback(context.Context) error {
+	return nil
+}
+
+// PreCheck makes sure that bootstrap phase is executed on the correct node
+func (p *updatePhaseBootstrapLeader) PreCheck(ctx context.Context) error {
+	return trace.Wrap(p.bootstrap.PreCheck(ctx))
+}
+
+// PostCheck is no-op for bootstrap phase
+func (p *updatePhaseBootstrapLeader) PostCheck(context.Context) error {
+	return nil
+}
+
+func (p *updatePhaseBootstrapLeader) rotateConfigAndSecrets() error {
+	for _, server := range p.servers {
+		if server.Runtime.Update != nil {
+			if err := p.rotateSecrets(server); err != nil {
+				return trace.Wrap(err, "failed to rotate secrets for %v", server)
+			}
+			if err := p.rotatePlanetConfig(server); err != nil {
+				return trace.Wrap(err, "failed to rotate planet configuration for %v", server)
+			}
+		}
+		if server.Teleport.Update != nil {
+			if err := p.rotateTeleportConfig(server); err != nil {
+				return trace.Wrap(err, "failed to rotate teleport configuration for %v", server)
+			}
+		}
+	}
+	return nil
+}
+
+func (p *updatePhaseBootstrapLeader) rotateSecrets(server storage.UpdateServer) error {
+	p.Infof("Generate new secrets configuration package for %v.", server)
+	resp, err := p.packageRotator.RotateSecrets(ops.RotateSecretsRequest{
+		Key:            p.bootstrap.Operation.ClusterKey(),
+		Package:        server.Runtime.SecretsPackage,
+		RuntimePackage: server.Runtime.Update.Package,
+		Server:         server.Server,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = p.bootstrap.Packages.CreatePackage(resp.Locator, resp.Reader, pack.WithLabels(resp.Labels))
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err)
+	}
+	p.Debugf("Rotated secrets package for %v: %v.", server, resp.Locator)
+	return nil
+}
+
+func (p *updatePhaseBootstrapLeader) rotatePlanetConfig(server storage.UpdateServer) error {
+	p.Infof("Generate new runtime configuration package for %v.", server)
+	resp, err := p.packageRotator.RotatePlanetConfig(ops.RotatePlanetConfigRequest{
+		Key:            p.bootstrap.Operation.Key(),
+		Server:         server.Server,
+		Manifest:       p.updateManifest,
+		RuntimePackage: server.Runtime.Update.Package,
+		Package:        &server.Runtime.Update.ConfigPackage,
+		Config:         p.existingClusterConfig,
+		Env:            p.existingEnviron,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = p.bootstrap.Packages.UpsertPackage(resp.Locator, resp.Reader,
```

Comment:
Is there any particular reason the planet config package is upserted, but the secrets package above is created (with already exists error ignored)?

Reply:
Hmm, I don't think there is any specific reason - I will update to use `UpsertPackage` API for the secrets too.